DllImportGenerator/DllImportGenerator/Analyzers/ManualTypeMarshallingAnalyzer.cs

```diff
@@ -343,101 +377,115 @@ private void AnalyzeNativeMarshalerType(SymbolAnalysisContext context, ITypeSymb
-                    }
-                }
-
-                if ((validateManagedGetPinnableReference && managedGetPinnableReferenceMethod is not null) 
-                    || ManualTypeMarshallingHelper.FindGetPinnableReference(marshalerType) is not null)
-                {
-                    // Validate that the Value property is a pointer-sized primitive type.
-                    if (valueProperty is null 
-                        || (valueProperty.Type is not (
-                            IPointerTypeSymbol _ or
-                            { SpecialType: SpecialType.System_IntPtr } or
-                            { SpecialType: SpecialType.System_UIntPtr })))
-                    {
-                        ITypeSymbol typeWithGetPinnableReference = managedGetPinnableReferenceMethod is not null
-                            ? type
-                            : marshalerType;
-
-                        context.ReportDiagnostic(Diagnostic.Create(NativeTypeMustBePointerSizedRule,
-                            valueProperty is not null
-                                ? GetSyntaxReferenceForDiagnostic(valueProperty).GetSyntax().GetLocation()
-                                : GetSyntaxReferenceForDiagnostic(nativeType).GetSyntax().GetLocation(),
-                            valuePropertyIsRefReturn
-                                ? $"ref {nativeType.ToDisplayString()}"
-                                : nativeType.ToDisplayString(),
-                            typeWithGetPinnableReference.ToDisplayString()));
+                        context.ReportDiagnostic(Diagnostic.Create(GetPinnableReferenceShouldSupportAllocatingMarshallingFallbackRule, 
+                            nativeMarshalerAttributeData.ApplicationSyntaxReference!.GetSyntax().GetLocation(),
+                            type.ToDisplayString()));
                     }
                 }
 
-                SyntaxReference GetSyntaxReferenceForDiagnostic(ISymbol targetSymbol)
+                if (getPinnableReferenceMethods is not (null, null) &&
+                    !valuePropertyIsRefReturn && // Ref returns are already reported above as invalid, so don't issue another warning here about them
+                    nativeType is not (
+                        IPointerTypeSymbol _ or
+                        { SpecialType: SpecialType.System_IntPtr } or
+                        { SpecialType: SpecialType.System_UIntPtr }))
                 {
-                    if (targetSymbol.DeclaringSyntaxReferences.IsEmpty)
+                    ITypeSymbol typeWithGetPinnableReference = getPinnableReferenceMethods switch
                     {
-                        return nativeMarshalerAttributeData.ApplicationSyntaxReference!;
-                    }
-                    else
-                    {
-                        return targetSymbol.DeclaringSyntaxReferences[0];
-                    }
+                        (IMethodSymbol managed, _) => managed.ContainingType,
```

Comment:
What about the non-null for both? Is that the same as the first case?

Reply:
Yes, non-null for both is the same as the first case.

Reply:
I actually thought the ternary was clearer than this switch / pattern match. That made it obvious that the logic was 'if the managed one is non-null, use that, otherwise use the marshaller one', whereas with the switch, I need to recognize that non-null for both should fall into the first case and the order matters.

Reply:
I switched to a null-coalescing expression to make it more explicit.